### PR TITLE
🧪 Introduce server-render folder #compiler

### PR DESCRIPTION
**Pull request by samouri**
**summary**

Creates a new top-level folder `server-render` for housing code to server-render AMP Documents, as well as a small skeleton of unimplemented functions to give y'all an idea of where my head is at and what I intend to build next.

File structure:
- `index.ts`: entrypoint. exports functions that transform an entire document.
- `ast.ts`: generic AST-related functionality. parsing, printing, converting back and forth between different formats.
- `render.ts`: for housing functions directly related to the rendering of AMP Components

Work for future PRs (to be discussed with wg-infra folks):
   - [ ] Ensure the `ts` is linted and prettified properly, as it isn't happening just yet.
   - [ ] Add explicit typechecking via `tsc` as part of `check-types`
   - [ ] Use a testing framework
   - [ ] Build as part of `amp dist|build`

### CODE REVIEW

**Review by rcebulko — COMMENTED** [review]
I was under the impression this was going to live in its own repo; what shifted it back here? (No strong opinion, just curious)

File: server-render/index.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import {parse, print} from './ast';
+import {renderAmpAst} from './render';
+
+/*
+ * Renders all of the eligible AMP Components within an AMP Document
+ */
+export function renderAmpHtml(html: string): string {
```
Comment (rcebulko) [inline server-render/index.ts:7]: nit: shouldn't this live in `./render` too?

**Comment by samouri** [comment]
> I was under the impression this was going to live in its own repo; what shifted it back here? (No strong opinion, just curious)

A few things:
1. It needs to directly import code from each component (either the Preact Component, or `buildDOM` function)
2. It needs to be built with the rest of the AMP in a versioned way.

Because of those two items, the only reasonable path I could think of is to include it in the `amphtml` repo.

**Comment by rcebulko** [comment]
> 2. It needs to be built with the rest of the AMP in a versioned way.

Could you clarify what this translates to specifically? ie. why couldn't one repo depend on the other through `package.json` and always have the latest version at build time?

**Comment by samouri** [comment]
> Could you clarify what this translates to specifically? ie. why couldn't one repo depend on the other through package.json and always have the latest version at build time?

We have infrastructure in place that builds AMP and places all of the RTVs code in both google3 and the CDN at a weekly cadence. If we wanted to accomplish this with a separate repo, we'd need to replicate so much of that. Am I overthinking the amount of infra work required?

**Comment by rcebulko** [comment]
I think you're right re: let all the end work happen in this repo. What I'm thinking is more that this can live in a separate repo, and then `amhtml` can have a dependency on it/use it to build and release the compiler each RTV. But the code itself (along with different style rules, Typescript/testing infra, etc) aren't bound to the `amphtml` conventions. Does that make sense?

File: server-render/index.ts (hunk 1 above)
Comment (samouri) [reply server-render/index.ts]: Done

**Review by rcebulko — APPROVED** [review]
I thought I approved this

**Comment by rcebulko** [comment]
> * [ ]  Add explicit typechecking via `tsc` as part of `check-types`

While `check-types` type-checks AMP source code, `check-build-system` is how the build-system code gets type-checked. I'm not sure if that means we should combine into one target, or keep type-check targets as separate tasks. WDYT?

**Pull request closed without merging**